### DISCUSSION TITLE
SSRCache not updating with nextjs getServerSideProps?

### DISCUSSION THREAD

**Discussion by cabbageofficial** [urql-graphql/urql#3045] [Q&A] [answered]
### Describe the bug

Hello, I am facing cache not updating issue, with nextjs.

Here, I fetch data into `getServerSideProps`-

```
export const getServerSideProps: GetServerSideProps = async (ctx) => {
    const { data } = await client.query(
        GET_STUDENT_LIST,
        { studentPrams: { name: "", studentId: "", class: "" } }, {
        fetchOptions: {
            headers: {
                cookie: ctx.req.headers.cookie as string,
                "user-agent": ctx.req.headers["user-agent"] as string,
                "x-forwarded-for": ctx.req.socket.remoteAddress as string
            },
        },
        requestPolicy: "network-only"
    }).toPromise();
    console.log("server data", data);
    return { props: { urqlState: ssrCache.extractData() } }
}
```

in my component I try to used cached data-

`const [{ data, error, fetching }, refetch] = useQuery<GetStudentListData>({ query: GET_STUDENT_LIST, variables: { studentPrams: { name: getValues("name") ?? "", studentId: getValues("studentId") ?? "", class: getValues("class") ?? "" } } });`

It worke perfectly when I first time to load my page, refresh my page. but If I change my route one page to another page, data do not updated properly, here is the video-

https://user-images.githubusercontent.com/88053948/225079984-b36025b7-24dd-4874-9c80-57f507b4ef91.mp4

here is the `client.ts`

```
import { createClient, ssrExchange, dedupExchange, cacheExchange, fetchExchange, errorExchange } from "urql";

const isServerSide = typeof window === "undefined";

const ssrCache = ssrExchange({ isClient: !isServerSide });

const client = createClient({
    url: "http://localhost:3001/ecampus",
    exchanges: [
        dedupExchange,
        errorExchange({
            onError: (error) => {
                error.message = error.message.replace(/^\[.*\]\s*/, "");
            }
        }),
        cacheExchange,
        ssrCache,
        fetchExchange

    ],
    fetchOptions: {
        credentials: "include"
    },
});

export { client, ssrCache };
```

and in `_app.tsx`-

```
import "@/styles/globals.css"
import type { AppProps } from "next/app";
import { Provider } from "urql";

//Graphql
import { client, ssrCache } from "@/Graphql/client";

//Fonts
import { inter } from "@/Fonts";

export default function App({ Component, pageProps }: AppProps) {
  if (pageProps.urqlState) {
    ssrCache.restoreData(pageProps.urqlState)
  }
  return (
    <Provider value={client}>
      <main className={`${inter.variable} font-sans`}>
        <Component {...pageProps} />
      </main>
    </Provider>
  )
}
```

Then what is the problem?

### Reproduction

https://github.com/cabbageofficial198/urql-problem

### Urql version

urql v3.0.3

### Validations

- [X] I can confirm that this is a bug report, and not a feature request, RFC, question, or discussion, for which [GitHub Discussions](https://github.com/urql-graphql/urql/discussions) should be used
- [X] Read the [docs](https://formidable.com/open-source/urql/docs/).
- [X] Follow our [Code of Conduct](https://github.com/urql-graphql/urql/blob/main/CODE_OF_CONDUCT.md)

**Comment by JoviDeCroock** [comment]
Converted to a discussion as it's a usage question it seems like 😅

**Reply by cabbageofficial** [reply]
@JoviDeCroock, I don't understand, can you explain it more?

**Reply by cabbageofficial** [reply]
@JoviDeCroock, please help me. I stacking here. I am not getting any solution. If you don't help me, I have no other way to continue my learning.

**Comment by kitten** [comment]
re. the replies above, these are really inappropriate. We're maintainers and while we help, we help in our own time voluntarily, so please don't just write this — not on GitHub, not to other maintainers, not on chats please.

I don't have time to currently run your example, but it seems you jump through multiple hoops to do what `next-urql` already does out of the box.

If you're using `getServerSideProps` with client-side fetching anyway, then you'll likely follow these instructions: https://formidable.com/open-source/urql/docs/advanced/server-side-rendering/#ssr-with--ssr-true-

**Comment by JoviDeCroock** [comment, marked as answer]
Now that I actually had a free moment to take a look at this, this is because of the defaultExchanges i.e. you are leveraging `[dedupExchange, cacheExchange, ssr, fetchExchange]` the `cacheExchange` is a document cache which has your data cached already so it returns a response and won't hit the `ssrExchange` here with the re-hydrated data.

[documentation on doc-cache](https://formidable.com/open-source/urql/docs/basics/document-caching/)

In this case the cache _should_ get invalidated by the `__typename` response of your mutation matching the one of your query, this is something for you to check

**Reply by cabbageofficial** [reply]
I can't figure out my problem. Would you please specify an example me. I see the docs, but everything going on top my head, I mean I am not understanding, what action I to do?

**Reply by JoviDeCroock** [reply]
I mean, you are making it a tad difficult for us to help you. Want to doubley call out that you are jumping hoops to avoid our `next-urql` package as called out in the other reply, that being said you would probably still run into this cache-thing.

Effort in open-source repositories preferably comes from both sides, hence having a shared example would maybe help us help you... currently you brush over it and say "I don't understand the docs" instead of telling us what you don't understand so we can improve as well.... Also calling out "please help me" is really not a great look 😅 we are doing this as volunteers as well...

In GraphQL every type has a `__typename` attribute which is used in the cache as explained in the docs.

Now what you need to check is two-fold, you check your list-query whether you see every entity there having `__typename: 'Student'` (or whatever other typename it has) and then check your mutation whether it has the _same_ value for its own `__typename`. If these match then the cache should invalidate the result of the query and hence let it through to your hydrated `ssrExchange`.

**Reply by cabbageofficial** [reply]
Yes, add additionalTypename for both useQuery and useMutation. It updating data perfectly, but it worked in csr. I mean data comes after render thhe page. 

But I use ssr data fetching. I can give code tomorrow. Actually my timezone it's midnight here.

**Reply by JoviDeCroock** [reply]
I wasn't really reaching for `additionalTypenames`, that would only be needed if no type can be derived this because you aren't asking for that data in your `selectionSet`, because your list is empty i.e. return value `[]` or because you are returning a scalar value.

The first step here was seeing why the data wasn't invalidated, also subsequent client invocations aren't really SSR. Your first paint is SSR, the subsequent navigations are Next sending some data over the wire which you choose to regard as part of the `ssr` data 😅 when we rehydrate that we add it to the `ssrExchange` meaning that when your query would hit it it would see the updated data coming from `getServerSideProps`, however we are seeing that the `cacheExchange` didn't evict yet, this could be because your list was empty i.e. it did not derive the type or because your mutation did not return the appropriate type to invalidate said cache.

**Reply by cabbageofficial** [reply]
I am giving my full code. You just say, where I have to change-

Here I have two nextjs pages-

`add-student.tsx` page-

In this page, I run a mutation by using `useMutation` hook on form submit. like-

`const [{ data, error, fetching }, addStudent] = useMutation<AddStudentData>(ADD_STUDENT);`

And `onSubmit` form handler-

```
const onSubmit: SubmitHandler<Inputs> = (value) => {
    const studentData = {
        ...value,
        dob: value.dob?.startDate,
        date: value.date?.startDate,
        image: "https://res.cloudinary.com/dub0dpenl/image/upload/v1678719134/Pie%20chat/S-30214_1_-_Copy_nhyp4x.jpg"
    };
    addStudent({ studentInput: studentData }).then(() => {
        setNotification(true)
    })
}
```

Here is `ADD_STUDENT` query-

```
export const ADD_STUDENT = gql`
mutation addStudent($studentInput: StudentInput!) {
    addStudent(studentInput: $studentInput) {
      success
      message
      __typename
    }
}
`;
```
Here `__typename` is "SuccessInfo" getting from backend

In this page I do not re-fetching student list as I not need student list in this page

and page no. 2 `student-list.tsx`-

In this page I am showing the student list. That's why I am fetching data in `getServerSideProps` 

```
export const getServerSideProps: GetServerSideProps = async (ctx) => {
    await client.query(
        GET_STUDENT_LIST,
        { studentPrams: { name: "", studentId: "", class: "" } }, {
        fetchOptions: {
            headers: {
                cookie: ctx.req.headers.cookie as string,
                "user-agent": ctx.req.headers["user-agent"] as string,
                "x-forwarded-for": ctx.req.socket.remoteAddress as string
            },
        },
        requestPolicy: "network-only",
    }).toPromise();
    return { props: { urqlState: ssrCache?.extractData() } }
}
```

And inner label component- I mean a list component into this page-

` const [{ data, error, fetching }, refetch] = useQuery<GetStudentListData>({ query: GET_STUDENT_LIST, variables: { studentPrams: { name:  "", studentId:  ?? "", class:  ?? "" } } });`

And use this data.

Here is `GET_STUDENT_LIST` query-

```
query getStudents($studentPrams: StudentPrams!) {
  getStudents(studentPrams: $studentPrams) {
    id
    studentId
    name
    fatherName
    dob
    number
    createdBy {
      name
      phone
    }
    __typename
  }
}
`;
```

Here `__typename` is "Student"

Now please say, where I have to change my code?

**Reply by JoviDeCroock** [reply]
I already see where this is wrong, your mutation isn't returning the `Student` type meaning your cache never knows what to invalidate you can fix this by giving `additionalTypenames` to your `mutation` _or_ by correcting this on your back-end. This is generally a best practise to return the entity you have operated on.

Another best practice in this case would be to prevent your empty list from not giving you a typename by doing a similar practice. These things are documented so I'm going to leave that detail be.

Good luck! :D

**Reply by cabbageofficial** [reply]
Now I can understand. Thank you very much..

**Comment by cabbageofficial** [comment]
@JoviDeCroock, I am now using next-urql in my project, with the documentations. But I am still getting the issue as show in my video before-

Now I am using `next-urql`-

```
export const getServerSideProps: GetServerSideProps = async (ctx) => {
    await client?.query(
        GET_CLASS_LIST,
        { searchInput: { search: "" } }, {
        fetchOptions: {
            headers: {
                cookie: ctx.req.headers.cookie as string,
                "user-agent": ctx.req.headers["user-agent"] as string,
                "x-forwarded-for": ctx.req.socket.remoteAddress as string
            },
        },
        requestPolicy: "network-only"
    }).toPromise();
    await client?.query(
        GET_STUDENT_LIST,
        { studentPrams: { name: "", studentId: "", class: "" } }, {
        fetchOptions: {
            headers: {
                cookie: ctx.req.headers.cookie as string,
                "user-agent": ctx.req.headers["user-agent"] as string,
                "x-forwarded-for": ctx.req.socket.remoteAddress as string
            },
        },
        requestPolicy: "network-only"
    }).toPromise();
    return { props: { urqlState: ssrCache?.extractData() } }
}

export default withUrqlClient(ssr => ({
    url: "http://localhost:3001/ecampus",
    fetchOptions: {
        credentials: "include"
    }
}))(StudentList);
```

And then in my component-

 ` const [{ data, error, fetching }, refetch] = useQuery<GetStudentListData>({ query: GET_STUDENT_LIST, variables: { studentPrams: { name: getValues("name") ?? "", studentId: getValues("studentId") ?? "", class: getValues("class") ?? "" } } });`

this is the video-

https://user-images.githubusercontent.com/88053948/225079984-b36025b7-24dd-4874-9c80-57f507b4ef91.mp4

But I am getting same problem. Can you please, tell me the where the problem occured here?

**Reply by JoviDeCroock** [reply]
Rather than tagging me, have you checked out the reply where I point out what is going wrong? https://github.com/urql-graphql/urql/discussions/3045#discussioncomment-5318510 😅 sorry but I am getting a tad annoyed with how you are treating our time that we dedicate for free

**Reply by cabbageofficial** [reply]
I follow your comment, I try to give same __typename, but not working at all?

**Reply by cabbageofficial** [reply]
@JoviDeCroock, Hello

**Comment by kitten** [comment]
There's not enough information here to help you, and we've tried to give you pointers. While I wish we had the time, even if we wanted to help you, if you say you've tried something, and we don't have any more information, then we can't provide more help and the outcome won't be the same.

Please don't `@`-tag people, especially maintainers. It's really irritating. We get notifications either way, and `@`-tagging just makes it impossible for us to mute them. In other words, it's like you called our phone each time you just wrote a single chat message.

Please debug this yourself for a bit. I can provide more help as needed, but only with more information.
That said, GitHub Discussions aren't a great format for this. It's not a chat, and if you basically demand fast/instant help from us, we _will_ unfortunately have to deem this a Code of Conduct violation.

Trust me, it's faster if you try debugging some things yourself. The alternative you're asking for here is that we reply to you instantly, however, that's like demanding our instant attention and time, so again, I will have to consider this a Code of Conduct violation if you continue to `@`-tag. Sorry

**Comment by cabbageofficial** [comment]
Okay. Sorry for disturbing you guys. I try debug it, I am getting proper solution from that debug as I am new and beginner.

Believe me, I  still facing the issues.

But okay, no problem. I have to stop my learning process here. Also, I will try a more bit into stack overflow.

After coming chatgpt, there are not people to place answer.

I have a question in my mind, If package maintainer or senior developer do not provide teaching to the junior, how can junior learn new things!

Okay, thank you guys.

**Reply by JoviDeCroock** [reply]
I think you are misunderstanding the point we are making...

Firstly, we are telling you that we can't help you if you just say "it didn't work", we don't know what you did so we can't help you any further as we are still basing ourselves off of the initial information you gave and we can't discover more.

Secondly, we are telling you that you shouldn't `@` us we will get back to you when we have time, we have jobs and don't have time on-demand to help you out, there are more issues/discussions and even repositories where we help out.

This isn't about a senior dev/... not wanting to help you, this is about you _DEMANDING_ our time which we have a limited amount off... When you ask for help it's extremely important that you share context and allow people to keep track with your discoveries rather than us having to think through everything you could do and having to assume that you performed our instructions correctly. Just mentioning this, as I called out that using @ and introducing new messages was annoying me and you immediately @'ed me again...

EDIT: also the new messages in a discussion are a bit hard to deal with, you can reply to messages and then we can actually have a conversation on what you're doing rather than what we're doing now

**Reply by cabbageofficial** [reply]
Okay, I can understand now. i should not use @.

Okay, then if I give you my repository link- please can you see it for finding the problem.

But there is one problem. My backend also running locally for testing you need also backend code.

Will I give you also my backend code?

**Reply by JoviDeCroock** [reply]
or instead reply to https://github.com/urql-graphql/urql/discussions/3045#discussioncomment-5318510 what the `__typename` properties are that you see for the query and mutation?
